### PR TITLE
pallet-collective: add event handler OnAbsentation

### PR DESCRIPTION
**Pull request by sorpaas**
This adds a new event handler `OnAbsentation` for `pallet-collective`. It is called when both of the following are true:

* The motion voting period has ended.
* A particular member of this collective did not vote.

This event handler is useful in implementing governance mechanics to deal with issues like council staling. For example, a blockchain can write the `OnAbsentation` event handler to be doing a call to `electionPhragmen::renounceCandidacy` -- if a council member did not vote on a motion within the whole voting period and caused a motion to stale with neither enough Ayes or Nays, then they're removed from council candidacies until they resubmit them, to create (more) incentives for council members to participate in voting.

polkadot companion: https://github.com/paritytech/polkadot/pull/1662

### CODE REVIEW

**Comment by shawntabrizi** [comment]
I think a PR like this makes _more_ sense if there was also an option to explicitly vote for neither side.

i.e. some sort of punishment on `OnAbsentation` basically forces users to now vote Aye/Nay, but gives no freedom to someone to truly has no side to an argument.

Would love to hear your thoughts on the matter. Otherwise, from an implementation standpoint, this looks fine.

**Comment by sorpaas** [comment]
@shawntabrizi That's a valid point, but right now a council member already cannot really vote "Abstain". If a council member does not vote, then the vote will default to prime member's vote. So even to "abstain" there will always be a valid vote for any council member. In particular, if a council member believes that more discussions are needed, then vote Nay, because a new motion can be created at any time.

**Comment by bkchr** [comment]
Any update?

**Comment by bkchr** [comment]
@sorpaas ?

**Comment by sorpaas** [comment]
@bkchr Let me fix the conflict! Other than that should be ready for review and merge.

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -659,17 +681,21 @@ decl_module! {
...
 				let proposal_count = Self::do_disapprove_proposal(proposal_hash);
-				return Ok(Some(
-					T::WeightInfo::close_disapproved(seats, proposal_count)
-				).into());
+
+				T::WeightInfo::close_disapproved(seats, proposal_count)
+			};
+
+			for member in absentation_members {
+				T::OnAbsentation::on_absentation(&member);
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:695]: The cost of on_absentation is not always part of the benchmark. for example the benchmark for close_disapproved have always less than 4 absentation.

Not sure how to this fix it though, maybe add some comment in the doc. or introduce some new benchmark.

**Review by thiolliere — COMMENTED** [review]
Logic is good to me, but I'm not sure benchmark cover potential new costs.

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (sorpaas) [reply frame/collective/src/lib.rs:695]: I think right now the benchmark is assuming `T::OnAbsentation` is `()`. For chains setting this to custom values, they probably need to provide their own `WeightInfo`.

Comment (thiolliere) [reply frame/collective/src/lib.rs:695]: benchmarks are run for the runtime, if user run benchmarks with a runtime where OnAbsentation is not no-op then the benchmark will run the logic of OnAbsentation. 

I think current strategy is to have user running benchmarks for their runtime, and just taking the resulting WeightInfo. Thus benchmarks should be written to cover all potential cost as far as possible.

Comment (thiolliere) [reply frame/collective/src/lib.rs:695]: though if benchmarking it is not easy I'm ok to just a comment somewhere saying that the benchmark doesn't handle the cost of OnAbsentation and that any costy OnAbsentation implementation must be careful and might need to assign weight manually.

**Comment by gnunicorn** [comment]
Closing due to lack of progress and being stale for a while.

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (sorpaas) [reply frame/collective/src/lib.rs:695]: > I'm ok to just a comment somewhere saying that the benchmark doesn't handle the cost of OnAbsentation and that any costy OnAbsentation implementation must be careful and might need to assign weight manually.

Sorry I missed your review on this. I'll add the comment.

**Comment by gnunicorn** [comment]
so,  what's the deal here?

**Comment by sorpaas** [comment]
@gnunicorn Ooops!

**Comment by sorpaas** [comment]
PR ready for review again. All current grumbles fixed!

**Review by thiolliere — APPROVED** [review]

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (thiolliere) [reply frame/collective/src/lib.rs:714]: ah I could have ping again also, all good now :-)

**Comment by sorpaas** [comment]
bot merge

**Review by shawntabrizi — COMMENTED** [review]
I dont see that this new pallet hook has proper weight integration.

**Comment by sorpaas** [comment]
I discussed with @shawntabrizi and the conclusion was that this PR is not ready because we haven't got proper benchmarking, due to the same issue @thiolliere mentioned earlier in the reivew. https://github.com/paritytech/substrate/pull/6985#pullrequestreview-489612570

Given this PR has been stale for so long, let me close it. Will come back and open a new PR when we figured out how to properly improve the benchmarking situations!

**Pull request closed without merging**